apps/vscode/src/providers/editor/toggle.ts

```diff
@@ -12,16 +12,99 @@
  * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
  *
  */
-
+import * as vscode from "vscode";
 import { commands, window, workspace, TextDocument, ViewColumn } from "vscode";
 import { Command } from "../../core/command";
 import { isQuartoDoc, kQuartoLanguageId } from "../../core/doc";
+import * as quarto from "quarto-core";
+import fs from "node:fs";
+import * as path from 'path';
+import yaml from "js-yaml";
 import { VisualEditorProvider } from "./editor";
 
 
+export async function determineMode(doc: TextDocument, config: string | undefined): Promise<boolean> {
+  const text = doc.getText()
+
+  let editorOpener = undefined;
+
+  // check if file itself has a mode
+  if (hasEditorMode(text, "source")) {
+    editorOpener = "source";
+  }
+  else if (hasEditorMode(text, "visual")) {
+    editorOpener = "visual";
+  }
+  // check if has a _quarto.yml or _quarto.yaml file with editor specified
+  else {
+    editorOpener = workspaceHasQuartoYaml();
+  }
+  if (editorOpener && editorOpener != config) {
+    editorOpener = editorOpener === 'visual' ? VisualEditorProvider.viewType : 'textEditor';
+    await vscode.commands.executeCommand('workbench.action.closeActiveEditor');
+    await vscode.commands.executeCommand("vscode.openWith",
+      doc.uri,
+      editorOpener
+    );
+    return true;
+  }
+
+  return false;
+}
+
+export async function setEditorOpener() {
+  const config = vscode.workspace.getConfiguration('quarto').get<string>('defaultEditor');
+  const viewType = config === 'visual' ? VisualEditorProvider.viewType : 'textEditor';
+  vscode.workspace.getConfiguration('workbench').update('editor.defaultView', viewType, true);
+  await vscode.commands.executeCommand("workbench.action.setDefaultEditor",
+    vscode.Uri.file('filename.qmd'),
+    viewType
+  );
+}
+
+export function workspaceHasQuartoYaml() {
```

Comment:
Maybe more like `modeFromQuartoYaml()`?

Reply:
Also specify return type? 🤔

Reply:
I'd be surprised if there wasn't already something like `getQuartoYaml()` somewhere else

Reply:
Maybe this?
https://github.com/quarto-dev/quarto/blob/97cc659a3481aca86b04e6c3fb5015d204d83ee8/packages/quarto-core/src/metadata.ts#L44

Combined with `yamlFromMetadata()`?
https://github.com/quarto-dev/quarto/blob/97cc659a3481aca86b04e6c3fb5015d204d83ee8/apps/vscode/src/providers/editor/prefs.ts#L213-L220